contracts/wallet/WalletFactory.sol

```diff
@@ -34,60 +38,161 @@ contract WalletFactory is Owned, Managed {
 
     // *************** Constructor ********************** //
 
     /**
      * @dev Default constructor.
      */
-    constructor(address _moduleRegistry, address _walletImplementation, address _ensManager) public {
+    constructor(
+        address _moduleRegistry,
+        address _walletImplementation,
+        address _ensManager,
+        address _guardianStorage
+    )
+        public
+    {
         moduleRegistry = _moduleRegistry;
         walletImplementation = _walletImplementation;
         ensManager = _ensManager;
+        guardianStorage = _guardianStorage;
     }
 
     // *************** External Functions ********************* //
 
     /**
-     * @dev Lets the manager create a wallet for an account. The wallet is initialised with a list of modules.
+     * @dev Lets the manager create a wallet for an owner account.
+     * The wallet is initialised with a list of modules and an optional ENS..
+     * The wallet is created using the CREATE opcode.
      * @param _owner The account address.
      * @param _modules The list of modules.
-     * @param _label Optional ENS label of the new wallet (e.g. franck).
+     * @param _label (Optional) ENS label of the new wallet, e.g. franck.
      */
-    function createWallet(address _owner, address[] calldata _modules, string calldata _label) external onlyManager {
-        require(_owner != address(0), "WF: owner cannot be null");
-        require(_modules.length > 0, "WF: cannot assign with less than 1 module");
-        require(ModuleRegistry(moduleRegistry).isRegisteredModule(_modules), "WF: one or more modules are not registered");
-        // create the proxy
+    function createWallet(
+        address _owner,
+        address[] calldata _modules,
+        string calldata _label
+    )
+        external
+        onlyManager
+    {
         Proxy proxy = new Proxy(walletImplementation);
         address payable wallet = address(proxy);
-        // check for ENS
-        bytes memory labelBytes = bytes(_label);
-        if (labelBytes.length != 0) {
-            // add the factory to the modules so it can claim the reverse ENS
-            address[] memory extendedModules = new address[](_modules.length + 1);
-            extendedModules[0] = address(this);
-            for(uint i = 0; i < _modules.length; i++) {
-                extendedModules[i + 1] = _modules[i];
-            }
-            // initialise the wallet with the owner and the extended modules
-            BaseWallet(wallet).init(_owner, extendedModules);
-            // register ENS
-            registerWalletENS(wallet, _label);
-            // remove the factory from the authorised modules
-            BaseWallet(wallet).authoriseModule(address(this), false);
-        } else {
-            // initialise the wallet with the owner and the modules
-            BaseWallet(wallet).init(_owner, _modules);
+        configureWallet(BaseWallet(wallet), _owner, _modules, _label, address(0));
```

Comment:
Just so I understand the change of logic between here and in `configureWallet`, why has the condition of having and ENS label changed? Also why does the WalletFactory need to be a temporarily registered module in the new wallet? And if it needs to be why does it have to be the first?

Reply:
The WalletFactory needs to be temporarily registered as a module so that:
- it can invoke the wallet and do the reverse ENS registration
- call the `GuardianStorage` contract and add the first guardian